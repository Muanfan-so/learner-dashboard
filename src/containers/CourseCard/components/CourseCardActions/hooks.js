import { handleEvent } from 'data/services/segment/utils';
import { eventNames } from 'data/services/segment/constants';

export const useCourseCardActionData = () => {
  const trackUpgradeClick = () => {
    handleEvent(eventNames.upgradeCourse, {
      pageName: 'learner_home',
      linkType: 'button',
      linkCategory: 'green_upgrade',
    });
  };

  return {
    trackUpgradeClick,
  };
};

export default useCourseCardActionData;
