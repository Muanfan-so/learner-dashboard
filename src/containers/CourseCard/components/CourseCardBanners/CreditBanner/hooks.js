import { StrictDict } from 'utils';
import { hooks as appHooks } from 'data/redux';

import ApprovedContent from './views/ApprovedContent';
import EligibleContent from './views/EligibleContent';
import MustRequestContent from './views/MustRequestContent';
import PendingContent from './views/PendingContent';
import RejectedContent from './views/RejectedContent';

export const statusComponents = StrictDict({
  pending: PendingContent,
  approved: ApprovedContent,
  rejected: RejectedContent,
});

export const useCreditBannerData = (cardId) => {
  const credit = appHooks.useCardCreditData(cardId);
  const { supportEmail } = appHooks.usePlatformSettingsData();
  if (!credit.isEligible) { return null; }

  const { error, purchased, requestStatus } = credit;
  let ContentComponent = EligibleContent;
  if (purchased) {
    if (requestStatus == null) {
      ContentComponent = MustRequestContent;
    } else if (Object.keys(statusComponents).includes(requestStatus)) {
      ContentComponent = statusComponents[requestStatus];
    }
    // Current behavior is to show Elligible State if unknown request status is returned
  }
  return {
    ContentComponent,
    error,
    supportEmail,
  };
};

export default {
  useCreditBannerData,
};
