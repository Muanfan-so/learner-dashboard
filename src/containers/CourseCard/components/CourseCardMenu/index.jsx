import React from 'react';
import PropTypes from 'prop-types';

import { Dropdown, Icon, IconButton } from '@edx/paragon';
import { MoreVert } from '@edx/paragon/icons';

import { hooks as appHooks } from 'data/redux';
import EmailSettingsModal from 'containers/EmailSettingsModal';
import UnenrollConfirmModal from 'containers/UnenrollConfirmModal';
import useCourseCardMenuData from './hooks';

export const CourseCardMenu = ({ cardId }) => {
  const {
    emailSettingsModal,
    unenrollModal,
  } = useCourseCardMenuData();
  const { isMasquerading } = appHooks.useMasqueradeData();
  return (
    <>
      <Dropdown>
        <Dropdown.Toggle
          id={`course-actions-dropdown-${cardId}`}
          as={IconButton}
          src={MoreVert}
          iconAs={Icon}
          variant="primary"
          alt="Actions dropdown"
        />
        <Dropdown.Menu>
          <Dropdown.Item disabled={isMasquerading} onClick={unenrollModal.show}>
            Unenroll
          </Dropdown.Item>
          <Dropdown.Item disabled={isMasquerading} onClick={emailSettingsModal.show}>
            Email Settings
          </Dropdown.Item>
          <Dropdown.Item href="#/action-3">Share to Facebook</Dropdown.Item>
          <Dropdown.Item href="#/action-3">Share to Twitter</Dropdown.Item>
        </Dropdown.Menu>
      </Dropdown>
      <UnenrollConfirmModal
        show={unenrollModal.isVisible}
        closeModal={unenrollModal.hide}
        cardId={cardId}
      />
      <EmailSettingsModal
        show={emailSettingsModal.isVisible}
        closeModal={emailSettingsModal.hide}
        cardId={cardId}
      />
    </>
  );
};
CourseCardMenu.propTypes = {
  cardId: PropTypes.string.isRequired,
};

export default CourseCardMenu;
