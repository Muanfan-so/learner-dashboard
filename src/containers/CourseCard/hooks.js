import { useIntl } from '@edx/frontend-platform/i18n';
import { hooks as appHooks } from 'data/redux';

export const useCardData = ({ cardId }) => {
  const { formatMessage } = useIntl();
  const { title, bannerImgSrc } = appHooks.useCardCourseData(cardId);
  const { isEnrolled } = appHooks.useCardEnrollmentData(cardId);

  return {
    isEnrolled,
    title,
    bannerImgSrc,
    formatMessage,
  };
};

export default useCardData;
