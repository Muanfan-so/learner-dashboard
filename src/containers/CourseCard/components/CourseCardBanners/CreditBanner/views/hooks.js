import React from 'react';

import { StrictDict } from 'utils';
import { apiHooks } from 'hooks';

import * as module from './hooks';

export const state = StrictDict({
  creditRequestData: (val) => React.useState(val), // eslint-disable-line
});

export const useCreditRequestData = (cardId) => {
  const [requestData, setRequestData] = module.state.creditRequestData(null);
  const createCreditApiRequest = apiHooks.useCreateCreditRequest(cardId);
  const createCreditRequest = (e) => {
    e.preventDefault();
    createCreditApiRequest().then(setRequestData);
  };
  return { requestData, createCreditRequest };
};

export default {
  useCreditRequestData,
};
