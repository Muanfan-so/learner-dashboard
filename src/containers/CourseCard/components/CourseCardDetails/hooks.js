import { useIntl } from '@edx/frontend-platform/i18n';
import { hooks as appHooks } from 'data/redux';

import * as hooks from './hooks';
import messages from './messages';

export const useAccessMessage = ({ cardId }) => {
  const { formatMessage, formatDate } = useIntl();
  const enrollment = appHooks.useCardEnrollmentData(cardId);
  const courseRun = appHooks.useCardCourseRunData(cardId);
  if (!courseRun.isStarted) {
    if (!courseRun.startDate) { return null; }
    const startDate = formatDate(courseRun.startDate);
    return formatMessage(messages.courseStarts, { startDate });
  }
  if (enrollment.isEnrolled) {
    const { isArchived, endDate } = courseRun;
    const {
      accessExpirationDate,
      isAuditAccessExpired,
    } = enrollment;

    if (enrollment.isAudit && accessExpirationDate) {
      return formatMessage(
        isAuditAccessExpired ? messages.accessExpired : messages.accessExpires,
        { accessExpirationDate: formatDate(accessExpirationDate) },
      );
    }
    if (!endDate) { return null; }
    return formatMessage(
      isArchived ? messages.courseEnded : messages.courseEnds,
      { endDate: formatDate(endDate) },
    );
  }
  return null;
};

export const useCardDetailsData = ({ dispatch, cardId }) => {
  const { formatMessage } = useIntl();
  const providerName = appHooks.useCardProviderData(cardId).name;
  const { courseNumber } = appHooks.useCardCourseData(cardId);
  const {
    isEntitlement,
    isFulfilled,
    canChange,
  } = appHooks.useCardEntitlementData(cardId);

  const openSessionModal = appHooks.useUpdateSelectSessionModalCallback(dispatch, cardId);

  return {
    providerName: providerName || formatMessage(messages.unknownProviderName),
    accessMessage: hooks.useAccessMessage({ cardId }),
    isEntitlement,
    isFulfilled,
    canChange,
    openSessionModal,
    courseNumber,
    changeOrLeaveSessionMessage: formatMessage(messages.changeOrLeaveSessionButton),
  };
};

export default useCardDetailsData;
