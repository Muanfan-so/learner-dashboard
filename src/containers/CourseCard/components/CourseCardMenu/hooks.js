import React from 'react';
import { StrictDict } from 'utils';

import track from 'tracking';
import { reduxHooks } from 'hooks';

import * as module from './hooks';

export const state = StrictDict({
  isUnenrollConfirmVisible: (val) => React.useState(val), // eslint-disable-line
  isEmailSettingsVisible: (val) => React.useState(val), // eslint-disable-line
});

export const useUnenrollData = () => {
  const [isVisible, setIsVisible] = module.state.isUnenrollConfirmVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const useEmailSettings = () => {
  const [isVisible, setIsVisible] = module.state.isEmailSettingsVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const useHandleToggleDropdown = (cardId) => {
  const eventName = track.course.courseOptionsDropdownClicked;
  const trackCourseEvent = reduxHooks.useTrackCourseEvent(eventName, cardId);
  return (isOpen) => {
    if (isOpen) { trackCourseEvent(); }
  };
};

export const useCourseCardMenu = (cardId) => {
  const { courseName } = reduxHooks.useCardCourseData(cardId);
  const { isEnrolled, isEmailEnabled } = reduxHooks.useCardEnrollmentData(cardId);
  const { twitter, facebook } = reduxHooks.useCardSocialSettingsData(cardId);
  const { isMasquerading } = reduxHooks.useMasqueradeData();
  const { isEarned } = reduxHooks.useCardCertificateData(cardId);
  const handleTwitterShare = reduxHooks.useTrackCourseEvent(
    track.socialShare,
    cardId,
    'twitter',
  );
  const handleFacebookShare = reduxHooks.useTrackCourseEvent(
    track.socialShare,
    cardId,
    'facebook',
  );

  const showUnenrollItem = isEnrolled && !isEarned;
  const showDropdown = showUnenrollItem || isEmailEnabled || facebook.isEnabled || twitter.isEnabled;

  return {
    courseName,
    isMasquerading,
    isEmailEnabled,
    showUnenrollItem,
    showDropdown,
    facebook,
    twitter,
    handleTwitterShare,
    handleFacebookShare,
  };
};
