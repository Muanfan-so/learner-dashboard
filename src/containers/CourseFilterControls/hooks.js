import React from 'react';
import { useToggle } from '@edx/paragon';

import { StrictDict } from 'utils';

import * as module from './hooks';

export const state = StrictDict({
  target: (val) => React.useState(val), // eslint-disable-line
});

export const useCourseFilterControlsData = ({
  setFilters,
  setSortBy,
}) => {
  const [isOpen, open, close] = useToggle(false);
  const [target, setTarget] = module.state.target(null);
  const handleFilterChange = ({ target: { checked, value } }) => {
    const update = checked ? setFilters.add : setFilters.remove;
    update(value);
  };
  const handleSortChange = ({ target: { value } }) => {
    setSortBy(value);
  };
  return {
    isOpen,
    open,
    close,
    target,
    setTarget,
    handleFilterChange,
    handleSortChange,
  };
};

export default useCourseFilterControlsData;
