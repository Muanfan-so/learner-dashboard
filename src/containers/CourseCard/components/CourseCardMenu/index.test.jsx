import { shallow } from 'enzyme';

import { reduxHooks } from 'hooks';
import { useEmailSettings, useUnenrollData } from './hooks';
import CourseCardMenu from '.';

jest.mock('react-share', () => ({
  FacebookShareButton: () => 'FacebookShareButton',
  TwitterShareButton: () => 'TwitterShareButton',
}));
jest.mock('hooks', () => ({
  reduxHooks: {
    useCardCourseData: jest.fn(),
    useCardEnrollmentData: jest.fn(),
    useCardSocialSettingsData: jest.fn(),
    useMasqueradeData: jest.fn(),
    useCardCertificateData: jest.fn(),
    useTrackCourseEvent: (_, __, site) => jest.fn().mockName(`${site}ShareClick`),
  },
}));
jest.mock('./hooks', () => ({
  useEmailSettings: jest.fn(),
  useUnenrollData: jest.fn(),
  useHandleToggleDropdown: jest.fn(),
}));

const props = {
  cardId: 'test-card-id',
};
const defaultEmailSettingsModal = {
  isVisible: false,
  show: jest.fn().mockName('emailSettingShow'),
  hide: jest.fn().mockName('emailSettingHide'),
};
const defaultUnenrollModal = {
  isVisible: false,
  show: jest.fn().mockName('unenrollShow'),
  hide: jest.fn().mockName('unenrollHide'),
};
const defaultSocialShare = {
  facebook: {
    isEnabled: true,
    shareUrl: 'facebook-share-url',
    socialBrand: 'facebook-social-brand',
  },
  twitter: {
    isEnabled: true,
    shareUrl: 'twitter-share-url',
    socialBrand: 'twitter-social-brand',
  },
};
const courseName = 'test-course-name';
let wrapper;
let el;

describe('CourseCardMenu', () => {
  const mockCourseCardMenu = ({
    isEnrolled,
    isEmailEnabled,
    isMasquerading,
    facebook,
    twitter,
    isEarned,
  }) => {
    useEmailSettings.mockReturnValueOnce(defaultEmailSettingsModal);
    useUnenrollData.mockReturnValueOnce(defaultUnenrollModal);
    reduxHooks.useCardCourseData.mockReturnValueOnce({ courseName });
    reduxHooks.useCardSocialSettingsData.mockReturnValueOnce({
      facebook: {
        ...defaultSocialShare.facebook,
        ...facebook,
      },
      twitter: {
        ...defaultSocialShare.twitter,
        ...twitter,
      },
    });
    reduxHooks.useCardEnrollmentData.mockReturnValueOnce({
      isEnrolled,
      isEmailEnabled,
    });
    reduxHooks.useMasqueradeData.mockReturnValueOnce({ isMasquerading });
    reduxHooks.useCardCertificateData.mockReturnValueOnce({ isEarned });
    return shallow(<CourseCardMenu {...props} />);
  };
  describe('enrolled, share enabled, email setting enable', () => {
    beforeEach(() => {
      wrapper = mockCourseCardMenu({
        isEnrolled: true,
        isEmailEnabled: true,
        isMasquerading: false,
        isEarned: false,
      });
    });
    test('snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('renders share buttons', () => {
      el = wrapper.find('FacebookShareButton');
      expect(el.length).toEqual(1);
      expect(el.prop('url')).toEqual('facebook-share-url');
      el = wrapper.find('TwitterShareButton');
      expect(el.length).toEqual(1);
      expect(el.prop('url')).toEqual('twitter-share-url');
    });
    it('renders enabled unenroll modal toggle', () => {
      el = wrapper.find({ 'data-testid': 'unenrollModalToggle' });
      expect(el.props().disabled).toEqual(false);
    });
    it('renders enabled email settings modal toggle', () => {
      el = wrapper.find({ 'data-testid': 'emailSettingsModalToggle' });
      expect(el.props().disabled).toEqual(false);
    });
    it('renders enabled email settings modal toggle', () => {
      el = wrapper.find({ 'data-testid': 'emailSettingsModalToggle' });
      expect(el.props().disabled).toEqual(false);
    });
  });
  describe('disable and stop rendering buttons', () => {
    it('does not render unenroll dropdown item when certificate is already earned', () => {
      wrapper = mockCourseCardMenu({
        isEnrolled: true,
        isEmailEnabled: true,
        isMasquerading: false,
        isEarned: true,
      });
      el = wrapper.find({ 'data-testid': 'unenrollModalToggle' });
      expect(el.length).toEqual(0);
    });
    it('does not render unenroll dropdown item when course is not enrolled', () => {
      wrapper = mockCourseCardMenu({
        isEnrolled: false,
        isEmailEnabled: true,
        isMasquerading: false,
        isEarned: false,
      });
      el = wrapper.find({ 'data-testid': 'unenrollModalToggle' });
      expect(el.length).toEqual(0);
    });
    it('does not render email settings modal toggle when email is not enabled', () => {
      wrapper = mockCourseCardMenu({
        isEnrolled: true,
        isEmailEnabled: false,
        isMasquerading: false,
        isEarned: false,
      });
      el = wrapper.find({ 'data-testid': 'emailSettingsModalToggle' });
      expect(el.length).toEqual(0);
    });
    it('does not render facebook share button when facebook is not enabled', () => {
      wrapper = mockCourseCardMenu({
        isEnrolled: true,
        isEmailEnabled: true,
        facebook: {
          ...defaultSocialShare.facebook,
          isEnabled: false,
        },
        isMasquerading: false,
        isEarned: false,
      });
      el = wrapper.find('FacebookShareButton');
      expect(el.length).toEqual(0);
    });
    it('does not render twitter share button when twitter is not enabled', () => {
      wrapper = mockCourseCardMenu({
        isEnrolled: true,
        isEmailEnabled: true,
        twitter: {
          ...defaultSocialShare.twitter,
          isEnabled: false,
        },
        isMasquerading: false,
        isEarned: false,
      });
      el = wrapper.find('TwitterShareButton');
      expect(el.length).toEqual(0);
    });
    it('snapshot when no dropdown items exist', () => {
      wrapper = mockCourseCardMenu({
        isEnrolled: false,
        isEmailEnabled: false,
        isMasquerading: false,
        isEarned: false,
        twitter: {
          ...defaultSocialShare.twitter,
          isEnabled: false,
        },
        facebook: {
          ...defaultSocialShare.facebook,
          isEnabled: false,
        },
      });
      expect(wrapper).toMatchSnapshot();
      expect(wrapper.isEmptyRender()).toEqual(true);
    });
  });
  describe('masquerading', () => {
    beforeEach(() => {
      wrapper = mockCourseCardMenu({
        isEnrolled: true,
        isEmailEnabled: true,
        isMasquerading: true,
        isEarned: false,
      });
    });
    test('snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });
    it('renders share buttons', () => {
      expect(wrapper.find('FacebookShareButton').length).toEqual(1);
      el = wrapper.find('TwitterShareButton');
      expect(el.length).toEqual(1);
      expect(el.prop('url')).toEqual('twitter-share-url');
    });
    it('renders disabled unenroll modal toggle', () => {
      el = wrapper.find({ 'data-testid': 'unenrollModalToggle' });
      expect(el.props().disabled).toEqual(true);
    });
    it('renders disabled email settings modal toggle', () => {
      el = wrapper.find({ 'data-testid': 'emailSettingsModalToggle' });
      expect(el.props().disabled).toEqual(true);
    });
  });
});
