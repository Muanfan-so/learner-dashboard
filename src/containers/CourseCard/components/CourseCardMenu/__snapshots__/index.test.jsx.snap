// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseCardMenu disable and stop rendering buttons snapshot when no dropdown items exist 1`] = `""`;

exports[`CourseCardMenu enrolled, share enabled, email setting enable snapshot 1`] = `
<Fragment>
  <Dropdown>
    <Dropdown.Toggle
      alt="Course actions dropdown"
      as="IconButton"
      iconAs="Icon"
      id="course-actions-dropdown-test-card-id"
      src={[MockFunction icons.MoreVert]}
      variant="primary"
    />
    <Dropdown.Menu>
      <Dropdown.Item
        data-testid="unenrollModalToggle"
        disabled={false}
        onClick={[MockFunction unenrollShow]}
      >
        Unenroll
      </Dropdown.Item>
      <Dropdown.Item
        data-testid="emailSettingsModalToggle"
        disabled={false}
        onClick={[MockFunction emailSettingShow]}
      >
        Email settings
      </Dropdown.Item>
      <FacebookShareButton
        className="pgn__dropdown-item dropdown-item"
        onClick={[MockFunction facebookShareClick]}
        resetButtonStyle={false}
        title="I'm taking test-course-name online with facebook-social-brand.  Check it out!"
        url="facebook-share-url"
      >
        Share to Facebook
      </FacebookShareButton>
      <TwitterShareButton
        className="pgn__dropdown-item dropdown-item"
        onClick={[MockFunction twitterShareClick]}
        resetButtonStyle={false}
        title="I'm taking test-course-name online with twitter-social-brand.  Check it out!"
        url="twitter-share-url"
      >
        Share to Twitter
      </TwitterShareButton>
    </Dropdown.Menu>
  </Dropdown>
  <UnenrollConfirmModal
    cardId="test-card-id"
    closeModal={[MockFunction unenrollHide]}
    show={false}
  />
  <EmailSettingsModal
    cardId="test-card-id"
    closeModal={[MockFunction emailSettingHide]}
    show={false}
  />
</Fragment>
`;

exports[`CourseCardMenu masquerading snapshot 1`] = `
<Fragment>
  <Dropdown>
    <Dropdown.Toggle
      alt="Course actions dropdown"
      as="IconButton"
      iconAs="Icon"
      id="course-actions-dropdown-test-card-id"
      src={[MockFunction icons.MoreVert]}
      variant="primary"
    />
    <Dropdown.Menu>
      <Dropdown.Item
        data-testid="unenrollModalToggle"
        disabled={true}
        onClick={[MockFunction unenrollShow]}
      >
        Unenroll
      </Dropdown.Item>
      <Dropdown.Item
        data-testid="emailSettingsModalToggle"
        disabled={true}
        onClick={[MockFunction emailSettingShow]}
      >
        Email settings
      </Dropdown.Item>
      <FacebookShareButton
        className="pgn__dropdown-item dropdown-item"
        onClick={[MockFunction facebookShareClick]}
        resetButtonStyle={false}
        title="I'm taking test-course-name online with facebook-social-brand.  Check it out!"
        url="facebook-share-url"
      >
        Share to Facebook
      </FacebookShareButton>
      <TwitterShareButton
        className="pgn__dropdown-item dropdown-item"
        onClick={[MockFunction twitterShareClick]}
        resetButtonStyle={false}
        title="I'm taking test-course-name online with twitter-social-brand.  Check it out!"
        url="twitter-share-url"
      >
        Share to Twitter
      </TwitterShareButton>
    </Dropdown.Menu>
  </Dropdown>
  <UnenrollConfirmModal
    cardId="test-card-id"
    closeModal={[MockFunction unenrollHide]}
    show={false}
  />
  <EmailSettingsModal
    cardId="test-card-id"
    closeModal={[MockFunction emailSettingHide]}
    show={false}
  />
</Fragment>
`;
